MinPlatformPkg/Test/TestPointStubDxe/TestPointStubDxeMm.c

```diff
@@ -241,6 +242,110 @@ PublishMmTestPoint (
   }
 }
 
+/**
+  Notification function of END_OF_DXE event group.
+
+  This is a notification function registered on END_OF_DXE event group.
+  When End of DXE is signalled we get the size of the PiSmmCommunicationRegionTable
+  to allocate a runtime buffer used for communicating the MM Testpoint results.
+  This requires the allocated pages to be unblocked for MM which must occur before
+  ReadyToLock.
+
+  @param[in] Event        Event whose notification function is being invoked.
+  @param[in] Context      Pointer to the notification function's context.
+
+**/
+VOID
+EFIAPI
+OnEndOfDxe (
+  IN EFI_EVENT  Event,
+  IN VOID       *Context
+  )
+{
+  EFI_STATUS                                          Status;
+  UINTN                                               CommSize;
+  EFI_MM_COMMUNICATION_PROTOCOL                      *MmCommunication;
+  UINTN                                               MinimalSizeNeeded;
+  EDKII_PI_SMM_COMMUNICATION_REGION_TABLE             *PiSmmCommunicationRegionTable;
+  UINT32                                              Index;
+  EFI_MEMORY_DESCRIPTOR                               *Entry;
+  UINTN                                               Size;
+
+  Status = gBS->LocateProtocol(&gEfiMmCommunicationProtocolGuid, NULL, (VOID **)&MmCommunication);
+  if (EFI_ERROR(Status)) {
+    DEBUG ((DEBUG_INFO, "MmiHandlerTestPoint: Locate MmCommunication protocol - %r\n", Status));
+    return ;
+  }
+
+  MinimalSizeNeeded = EFI_PAGE_SIZE;
+
+  Status = EfiGetSystemConfigurationTable(
+             &gEdkiiPiSmmCommunicationRegionTableGuid,
+             (VOID **)&PiSmmCommunicationRegionTable
+             );
+  if (EFI_ERROR(Status)) {
+    DEBUG ((DEBUG_INFO, "MmiHandlerTestPoint: Get PiSmmCommunicationRegionTable - %r\n", Status));
+    return ;
+  }
+  ASSERT(PiSmmCommunicationRegionTable != NULL);
+  Entry = (EFI_MEMORY_DESCRIPTOR *)(PiSmmCommunicationRegionTable + 1);
+  Size = 0;
+  for (Index = 0; Index < PiSmmCommunicationRegionTable->NumberOfEntries; Index++) {
+    if (Entry->Type == EfiConventionalMemory) {
+      Size = EFI_PAGES_TO_SIZE((UINTN)Entry->NumberOfPages);
+      if (Size >= MinimalSizeNeeded) {
+        break;
+      }
+    }
+    Entry = (EFI_MEMORY_DESCRIPTOR *)((UINT8 *)Entry + PiSmmCommunicationRegionTable->DescriptorSize);
+  }
+  ASSERT(Index < PiSmmCommunicationRegionTable->NumberOfEntries);
+
+  CommSize = OFFSET_OF (EFI_MM_COMMUNICATE_HEADER, Data) + sizeof(MMI_HANDLER_TEST_POINT_PARAMETER_GET_DATA_BY_OFFSET);
+
+  Size -= CommSize;
+  mMmCommBuffer = AllocateRuntimeZeroPool(Size);

```

Comment:
It looked like the code above was searching `PiSmmCommunicationRegionTable` for a suitably sized comm buffer to use. This code allocates a new buffer though. Was that the goal? Normally, after finding a comm buffer of an acceptable size `Entry->PhysicalStart` would be used as the comm buffer.

Reply:
We want to allocate a new buffer.  It's very vaguely explained in the function description but in the original implementation we're using the comm buffer to transmit data from MM to the DXE so we can dump it in the DXE dump app.  Standalone MM adds additional checks, one for the comm buffer and another one to check if we're overlapping with MMRAM.  When we were in the Comm buffer region we weren't in user space and thus causing the second bound check to fail.  The memory here can be unblocked so that it can be accessed but this needs to happen before ReadyToLock.  I went around this by creating a new buffer that I was able to unblock before ReadyToLock that we can reference for this purpose as we can't do that with the original table.

Reply:
I kept most of the logic to make sure we're getting the correct table size when allocating a buffer.  I think all we need is a single page and I can test that tomorrow to see if the logic can be simplified.

Reply:
I completely reworked the logic to not have this buffer allocation so I'm closing this thread.